pkg/kv/kvserver/replica_raft_overload.go

```diff
@@ -206,3 +206,74 @@ func (osm *overloadedStoresMap) Swap(
 	v, _ := (*atomic.Value)(osm).Swap(m).(map[roachpb.StoreID]*admissionpb.IOThreshold)
 	return v
 }
+
+func (r *Replica) updatePausedFollowersLocked(
+	ctx context.Context, ioOverloadMap map[roachpb.StoreID]*admissionpb.IOThreshold,
+) {
+	r.mu.pausedFollowers = nil
+
+	if len(ioOverloadMap) == 0 {
+		return
+	}
+
+	if r.replicaID != r.mu.leaderID {
+		// Only the raft leader pauses followers. Followers never send meaningful
+		// amounts of data in raft messages, so pausing doesn't make sense on them.
+		return
+	}
+
+	if !quotaPoolEnabledForRange(*r.descRLocked()) {
+		// If the quota pool isn't enabled (like for the liveness range), play it
+		// safe. The range is unlikely to be a major contributor to any follower's
+		// I/O and wish to reduce the likelihood of a problem in replication pausing
+		// contributing to an outage of that critical range.
+		return
+	}
+
+	status := r.leaseStatusAtRLocked(ctx, r.Clock().NowAsClockTimestamp())
+	if !status.IsValid() || !status.OwnedBy(r.StoreID()) {
+		// If we're not the leaseholder (which includes the case in which we just
+		// transferred the lease away), leave all followers unpaused. Otherwise, the
+		// leaseholder won't learn that the entries it submitted were committed
+		// which effectively causes range unavailability.
+		return
+	}
```

Comment:
Unpausing if there's no valid lease seems like a prudent failsafe.

However, it seems a bit lenient to unpause all followers when we don't have the lease. There might be cases where this situation persists, or where it causes intermittent unpausing. Should we only exclude the leaseholder instead, or are we worried that might open for other pathologies?

Reply:
Really hard to say without the benefit of hindsight. But I'd much rather approach pausing from below (adding complexity only when proven relevant) vs trying to anticipate problems and then (perhaps painfully) finding out that there was something we missed.
I expect it to be easier to see that a node is still sending to an overloaded follower (esp. in light of the new admission control logs that are landing ~right about now) vs figuring out that some loss of quorum or other degradation is caused by pausing.